File: source/modules/strategy/src/test/sample_data/inputs/Positions.java
import java.math.BigDecimal;
import java.util.Collection;
import java.util.Date;
import java.util.Map;

import org.marketcetera.core.position.PositionKey;
import org.marketcetera.strategy.java.Strategy;
import org.marketcetera.trade.Equity;
import org.marketcetera.trade.Option;
import org.marketcetera.trade.OptionType;

/* $License$ */

/**
 * Tests the position-related API of the Strategy API.
 *
 * @author <a href="mailto:colin@marketcetera.com">Colin DuPlantis</a>
 * @version $Id$
 * @since 2.0.0
 */
public class Positions
        extends Strategy
{
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onStart()
     */
    @Override
    public void onStart()
    {
        doGetPositionAsOf(true);
        doGetAllPositionsAsOf(true);
        doGetOptionPositionAsOf(true);
        doGetAllOptionPositionsAsOf(true);
        doGetOptionPositionsAsOf(true);
        doGetUnderlying(true);
        doGetOptionRoots(true);
    }
    /* (non-Javadoc)
     * @see org.marketcetera.strategy.java.Strategy#onStop()
     */
    @Override
    public void onStop()
    {
        doGetPositionAsOf(false);
        doGetAllPositionsAsOf(false);
        doGetOptionPositionAsOf(false);
        doGetAllOptionPositionsAsOf(false);
        doGetOptionPositionsAsOf(false);
        doGetUnderlying(false);
        doGetOptionRoots(false);
    }
    /**
     * Executes {@link #getPositionAsOf(java.util.Date, String)}.
     */
    private void doGetPositionAsOf(boolean duringStart)
    {
        String symbol = getProperty("symbol");
        String dateString = getProperty("date");
        Date date;
        if(dateString == null) {
            date = null;
        } else {
            date = new Date(Long.parseLong(dateString));
        }
        BigDecimal result = getPositionAsOf(date,
                                            symbol);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("positionAsOf",
                        resultString);
        } else {
            setProperty("positionAsOfDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getAllPositionsAsOf(java.util.Date)}.
     */
    private void doGetAllPositionsAsOf(boolean duringStart)
    {
        String dateString = getProperty("date");
        Date date;
        if(dateString == null) {
            date = null;
        } else {
            date = new Date(Long.parseLong(dateString));
        }
        Map<PositionKey<Equity>,BigDecimal> result = getAllPositionsAsOf(date);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("allPositionsAsOf",
                        resultString);
        } else {
            setProperty("allPositionsAsOfDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getOptionPositionAsOf(java.util.Date, String, String, java.math.BigDecimal, org.marketcetera.trade.OptionType)}
     */
    private void doGetOptionPositionAsOf(boolean duringStart)
    {
        String optionRoot = getProperty("optionRoot");
        String expiry = getProperty("expiry");
        String strikePriceString = getProperty("strikePrice");
        BigDecimal strikePrice = null;
        if(strikePriceString != null) {
            strikePrice = new BigDecimal(strikePriceString);
        }
        String optionTypeString = getProperty("optionType");
        OptionType optionType = null;
        if(optionTypeString != null) {
            optionType = OptionType.valueOf(optionTypeString);
        }
        String dateString = getProperty("date");
        Date date = null;
        if(dateString != null) {
            date = new Date(Long.parseLong(dateString));
        }
        BigDecimal result = getOptionPositionAsOf(date,
                                                  optionRoot,
                                                  expiry,
                                                  strikePrice,
                                                  optionType);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("optionPositionAsOf",
                        resultString);
        } else {
            setProperty("optionPositionAsOfDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getAllOptionPositionsAsOf(java.util.Date)}.
     */
    private void doGetAllOptionPositionsAsOf(boolean duringStart)
    {
        String dateString = getProperty("date");
        Date date;
        if(dateString == null) {
            date = null;
        } else {
            date = new Date(Long.parseLong(dateString));
        }
        Map<PositionKey<Option>,BigDecimal> result = getAllOptionPositionsAsOf(date);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("allOptionPositionsAsOf",
                        resultString);
        } else {
            setProperty("allOptionPositionsAsOfDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getOptionPositionsAsOf(java.util.Date, String...)}.
     */
    private void doGetOptionPositionsAsOf(boolean duringStart)
    {
        String optionRootsString = getProperty("optionRoots");
        String[] optionRoots = null;
        if(optionRootsString != null &&
                !optionRootsString.isEmpty()) {
            optionRoots = optionRootsString.split(",");
        }
        String dateString = getProperty("date");
        Date date = null;
        if(dateString != null) {
            date = new Date(Long.parseLong(dateString));
        }
        Map<PositionKey<Option>,BigDecimal> result;
        if(getParameter("nullOptionRoot") != null) {
            result = getOptionPositionsAsOf(date,
                                            new String[] { optionRootsString, null } );
        } else if(getParameter("emptyOptionRoot") != null) {
            result = getOptionPositionsAsOf(date,
                                            new String[] { optionRootsString, "" } );
        } else {
            result = getOptionPositionsAsOf(date,
                                            optionRoots);
        }
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("optionPositionsAsOf",
                        resultString);
        } else {
            setProperty("optionPositionsAsOfDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getUnderlying(String)}. 
     */
    private void doGetUnderlying(boolean duringStart)
    {
        String optionRoot = getProperty("optionRoot");
        String result = getUnderlying(optionRoot);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("underlying",
                        resultString);
        } else {
            setProperty("underlyingDuringStop",
                        resultString);
        }
    }
    /**
     * Executes {@link #getOptionRoots(String)}.
     */
    private void doGetOptionRoots(boolean duringStart)
    {
        String underlyingSymbol = getProperty("underlyingSymbol");
        Collection<String> result = getOptionRoots(underlyingSymbol);
        String resultString;
        if(result == null) {
            resultString = null;
        } else {
            resultString = result.toString();
        }
        if(duringStart) {
            setProperty("optionRoots",
                        resultString);
        } else {
            setProperty("optionRootsDuringStop",
                        resultString);
        }
    }
}

File: root/tags/2.1.2/source/modules/strategy/src/test/sample_data/inputs/positions.rb
include_class "org.marketcetera.strategy.ruby.Strategy"
include_class "org.marketcetera.trade.OptionType"
include_class "java.lang.Long"
include_class "java.lang.String"
include_class "java.math.BigDecimal"
include_class "java.util.Date"

########################################
# Tests Strategy API positions methods #
########################################
class Positions < Strategy
  
    ##########################################
    # Executed when the strategy is started. #
    #                                        #
    # Use this method to set up data flows   #
    #  and other initialization tasks.       #
    ##########################################
    def on_start
        do_get_position_as_of true
        do_get_all_positions_as_of true
        do_get_option_position_as_of true
        do_get_all_option_positions_as_of true
        do_get_option_positions_as_of true
        do_get_underlying true
        do_get_option_roots true
    end
    
    ############################################
    # Executed when the strategy is stopped.   #
    ############################################
    def on_stop
        do_get_position_as_of false
        do_get_all_positions_as_of false
        do_get_option_position_as_of false
        do_get_all_option_positions_as_of false
        do_get_option_positions_as_of false
        do_get_underlying false
        do_get_option_roots false
    end
    ###############################
    # Executes get_position_as_of #
    ###############################
    def do_get_position_as_of(duringStart)
      symbol = get_property "symbol"
      dateString = get_property "date"
      if dateString.nil?
          date = nil
      else
          date = Date.new Long.parseLong dateString
      end
      result = get_position_as_of date, symbol
      if result.nil?
          resultString = nil
      else
          resultString = result.to_s
      end
      if duringStart
          set_property "positionAsOf", resultString
      else
          set_property "positionAsOfDuringStop", resultString
      end
    end
    ####################################
    # Executes get_all_positions_as_of #
    ####################################
    def do_get_all_positions_as_of(duringStart)
        dateString = get_property "date"
        if dateString.nil?
            date = nil
        else
            date = Date.new Long.parseLong dateString
        end
        result = get_all_positions_as_of date
        if result.nil?
            resultString = nil
        else
            resultString = result.to_s
        end
        if duringStart
            set_property "allPositionsAsOf", resultString
        else
            set_property "allPositionsAsOfDuringStop", resultString
        end
    end
    ######################################
    # Executes get_option_position_as_of #
    ######################################
    def do_get_option_position_as_of(duringStart)
        optionRoot = get_property "optionRoot"
        expiry = get_property "expiry"
        strikePriceString = get_property "strikePrice"
        strikePrice = nil
        if !strikePriceString.nil?
            strikePrice = BigDecimal.new strikePriceString
        end
        optionTypeString = get_property "optionType"
        optionType = nil
        if !optionTypeString.nil?
            optionType = OptionType.valueOf optionTypeString
        end
        dateString = get_property "date"
        date = nil
        if !dateString.nil?
            date = Date.new Long.parseLong dateString
        end
        result = get_option_position_as_of date,optionRoot,expiry,strikePrice,optionType
        if result.nil?
            resultString = nil
        else
            resultString = result.to_s
        end
        if duringStart
            set_property "optionPositionAsOf",resultString
        else
            set_property "optionPositionAsOfDuringStop",resultString
        end
    end
    ###########################################
    # Executes get_all_option_positions_as_of #
    ###########################################
    def do_get_all_option_positions_as_of(duringStart)
        dateString = get_property "date"
        date = nil
        if !dateString.nil?
            date = Date.new Long.parseLong dateString
        end
        result = get_all_option_positions_as_of date
        if result.nil?
            resultString = nil
        else
            resultString = result.to_s
        end
        if duringStart
            set_property "allOptionPositionsAsOf",resultString
        else
            set_property "allOptionPositionsAsOfDuringStop",resultString
        end
    end
    #######################################
    # Executes get_option_positions_as_of #
    #######################################
    def do_get_option_positions_as_of(duringStart)
        optionRootsString = get_property "optionRoots"
        optionRoots = nil
        if !optionRootsString.nil? && !optionRootsString.empty?
            optionRoots = optionRootsString.split ","
        end
        dateString = get_property "date"
        date = nil
        if !dateString.nil?
            date = Date.new Long.parseLong dateString
        end
        if !get_parameter("nullOptionRoot").nil?
            result = get_option_positions_as_of date, (Array.new << optionRootsString << nil).to_java(String)
        else
            if !get_parameter("emptyOptionRoot").nil?
                result = get_option_positions_as_of date, (Array.new << optionRootsString << "").to_java(String)
            else
                result = get_option_positions_as_of date, (optionRoots.nil? ? nil : optionRoots.to_java(String))
            end
        end
        if result.nil?
            resultString = nil
        else
           resultString = result.to_s
        end
        if duringStart
            set_property "optionPositionsAsOf",resultString
        else
            set_property "optionPositionsAsOfDuringStop",resultString
        end
    end
    ###########################
    # Executes get_underlying #
    ###########################
    def do_get_underlying(duringStart)
        optionRoot = get_property "optionRoot"
        result = get_underlying optionRoot
        if result.nil?
            resultString = nil
        else
           resultString = result.to_s
        end
        if duringStart
            set_property "underlying",resultString
        else
            set_property "underlyingDuringStop",resultString
        end
    end
    #############################
    # Executes get_option_roots #
    #############################
    def do_get_option_roots(duringStart)
        underlyingSymbol = get_property "underlyingSymbol"
        result = get_option_roots underlyingSymbol
        if result.nil?
            resultString = nil
        else
            resultString = result.to_s
        end
        if duringStart
            set_property "optionRoots",resultString
        else
            set_property "optionRootsDuringStop",resultString
        end
    end
end
